opentelemetry-sdk/src/trace/mod.rs

```diff
@@ -76,6 +76,49 @@ mod tests {
         Context, KeyValue,
     };
 
+    #[test]
+    fn span_context_immutable() {
+        let exporter = InMemorySpanExporterBuilder::new().build();
+        let provider = SdkTracerProvider::builder()
+            .with_span_processor(SimpleSpanProcessor::new(exporter.clone()))
+            .build();
+        let tracer = provider.tracer("test_tracer");
+
+        #[derive(Debug, PartialEq)]
+        struct ValueA(u64);
+
+        let span = tracer.start("span-name");
+
+        // start with Current, which should have no span
+        let cx = Context::current();
+        assert!(cx.has_active_span() == false);
+
+        // add span to context
+        let cx_with_span = cx.with_span(span);
+        assert!(cx_with_span.has_active_span() == true);
+        assert!(cx.has_active_span() == false);
+
+        // modify the span in the context
+        let span_ref = cx_with_span.span();
+        span_ref.set_attribute(KeyValue::new("attribute1", "value1"));
+
+        // create a new context, which should not affect the original
+        let cx_with_span_and_more = cx_with_span.with_value(ValueA(1));
+
+        // modify the span in the new context, which should not affect the original
```

Comment:
but it affects the original, as the new Context is Arc pointer to the original Span itself...

Reply:
Isn't this how it's supposed to work? I don't understand what you think is wrong here. If you do something to a `Span` that should be reflected on the `Span` right? The `Span` API has explicitly mutating operations like `set_attribute`, and they should change the `Span`.

The `Context` being **immutable** doesn't mean that all things in the `Context` are **immutable**.

Reply:
Yes I expect Span modification to take effect on the Span, but the Context immutability part confused me. I'll modify the test so that it validates the expected behavior.